pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/BaseBooleanAggregateFunction.java

```diff
@@ -0,0 +1,249 @@
+import org.apache.pinot.core.query.aggregation.groupby.ObjectGroupByResultHolder;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.roaringbitmap.RoaringBitmap;
+
+
+public abstract class BaseBooleanAggregateFunction extends BaseSingleInputAggregationFunction<Integer, Integer> {
+
+  private final BooleanMerge _merger;
+  private final boolean _nullHandlingEnabled;
+
+  protected enum BooleanMerge {
+    AND {
+      @Override
+      long merge(long left, int right) {
+        return left * right;
+      }
+
+      @Override
+      boolean isTerminal(long agg) {
+        return agg == 0;
+      }
+
+      @Override
+      long getDefaultValue() {
+        return 1;
+      }
+    },
+    OR {
+      @Override
+      long merge(long left, int right) {
+        return left + right;
+      }
+
+      @Override
+      boolean isTerminal(long agg) {
+        return agg > 0;
+      }
+
+      @Override
+      long getDefaultValue() {
+        return 0;
+      }
+    };
+
+    abstract long merge(long left, int right);
+
+    abstract boolean isTerminal(long agg);
+
+    abstract long getDefaultValue();
+  }
+
+  protected BaseBooleanAggregateFunction(ExpressionContext expression, boolean nullHandlingEnabled,
+      BooleanMerge merger) {
+    super(expression);
+    _nullHandlingEnabled = nullHandlingEnabled;
+    _merger = merger;
+  }
+
+  @Override
+  public AggregationResultHolder createAggregationResultHolder() {
+    return _nullHandlingEnabled
+        ? new ObjectAggregationResultHolder()
+        : new LongAggregationResultHolder(_merger.getDefaultValue());
+  }
+
+  @Override
+  public GroupByResultHolder createGroupByResultHolder(int initialCapacity, int maxCapacity) {
+    return _nullHandlingEnabled
+        ? new ObjectGroupByResultHolder(initialCapacity, maxCapacity)
+        : new LongGroupByResultHolder(initialCapacity, maxCapacity, _merger.getDefaultValue());
+  }
+
+  @Override
+  public void aggregate(int length, AggregationResultHolder aggregationResultHolder,
+      Map<ExpressionContext, BlockValSet> blockValSetMap) {
+    BlockValSet blockValSet = blockValSetMap.get(_expression);
+
+    if (blockValSet.getValueType().getStoredType() != FieldSpec.DataType.INT) {
+      throw new IllegalArgumentException(
+          String.format("Unsupported data type %s for BOOL_AND", blockValSet.getValueType()));
+    }
+
+    int[] bools = blockValSet.getIntValuesSV();
+    if (_nullHandlingEnabled) {
+      long agg = getLong(aggregationResultHolder.getResult());
+
+      RoaringBitmap nullBitmap = blockValSet.getNullBitmap();
+      if (nullBitmap == null) {
+        nullBitmap = new RoaringBitmap();
+      } else if (nullBitmap.getCardinality() > length) {
+        return;
+      }
+
+      for (int i = 0; i < length; i++) {
+        if (!nullBitmap.contains(i)) {
+          agg = _merger.merge(agg, bools[i]);
+          aggregationResultHolder.setValue((Object) agg);
+          if (_merger.isTerminal(agg)) {
```

Comment:
Move this check to line 117 so that the block can be skipped.
We might not want to do per-value terminal check because tight loop tends to have better performance. We can do early terminate on block level